File: backend/app/routers/agents.py
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, HTTPException, Request
from sqlmodel import select
from app.db import get_session
from app.models import Agent, AgentSession, TaskAssignment, AgentMessage
from app.schemas import (
    AgentCreate,
    AgentUpdate,
    AgentRead,
    AgentSessionCreate,
    AgentSessionRead,
    TaskAssignmentCreate,
    TaskAssignmentUpdate,
    TaskAssignmentRead,
    AgentMessageSend,
    AgentMessageRead,
)

router = APIRouter(prefix="/agents", tags=["agents"])

def _agent_payload(agent: Agent) -> dict:
    return {
        "id": agent.id,
        "name": agent.name,
        "capabilities": agent.capabilities,
        "status": agent.status,
        "metadata": agent.agent_metadata,
        "created_at": agent.created_at,
        "updated_at": agent.updated_at,
    }

def _message_payload(message: AgentMessage) -> dict:
    return {
        "id": message.id,
        "from_agent_id": message.from_agent_id,
        "to_agent_id": message.to_agent_id,
        "content": message.content,
        "message_type": message.message_type,
        "task_id": message.task_id,
        "read": message.read,
        "created_at": message.created_at,
    }


@router.post("", response_model=AgentRead)
def create_agent(payload: AgentCreate):
    with get_session() as session:
        agent = Agent(**payload.dict())
        session.add(agent)
        session.commit()
        session.refresh(agent)
        return _agent_payload(agent)


@router.get("", response_model=list[AgentRead])
def list_agents(status: Optional[str] = None):
    with get_session() as session:
        stmt = select(Agent)
        if status is not None:
            stmt = stmt.where(Agent.status == status)
        agents = session.exec(stmt.order_by(Agent.updated_at.desc())).all()
        return [_agent_payload(agent) for agent in agents]


@router.get("/{agent_id}", response_model=AgentRead)
def get_agent(agent_id: int):
    with get_session() as session:
        agent = session.get(Agent, agent_id)
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        return _agent_payload(agent)


@router.patch("/{agent_id}", response_model=AgentRead)
def update_agent(agent_id: int, payload: AgentUpdate):
    with get_session() as session:
        agent = session.get(Agent, agent_id)
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        for k, v in payload.dict(exclude_unset=True).items():
            setattr(agent, k, v)
        agent.updated_at = datetime.utcnow()
        session.add(agent)
        session.commit()
        session.refresh(agent)
        return _agent_payload(agent)


@router.post("/{agent_id}/sessions", response_model=AgentSessionRead)
def start_session(agent_id: int, payload: AgentSessionCreate):
    with get_session() as session:
        agent = session.get(Agent, agent_id)
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        session_obj = AgentSession(
            agent_id=agent_id,
            context=payload.context,
        )
        agent.status = "online"
        session.add(session_obj)
        session.add(agent)
        session.commit()
        session.refresh(session_obj)
        return session_obj


@router.get("/{agent_id}/sessions", response_model=list[AgentSessionRead])
def list_sessions(agent_id: int):
    with get_session() as session:
        sessions = session.exec(
            select(AgentSession)
            .where(AgentSession.agent_id == agent_id)
            .order_by(AgentSession.started_at.desc())
        ).all()
        return sessions


@router.post("/{agent_id}/sessions/{session_id}/end", response_model=AgentSessionRead)
def end_session(agent_id: int, session_id: int):
    with get_session() as session:
        agent = session.get(Agent, agent_id)
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        session_obj = session.get(AgentSession, session_id)
        if not session_obj or session_obj.agent_id != agent_id:
            raise HTTPException(status_code=404, detail="Session not found")
        session_obj.ended_at = datetime.utcnow()
        agent.status = "offline"
        session.add(session_obj)
        session.add(agent)
        session.commit()
        session.refresh(session_obj)
        return session_obj


@router.post("/assignments", response_model=TaskAssignmentRead)
def create_assignment(payload: TaskAssignmentCreate):
    with get_session() as session:
        assignment = TaskAssignment(**payload.dict())
        session.add(assignment)
        session.commit()
        session.refresh(assignment)
        return assignment


@router.get("/assignments", response_model=list[TaskAssignmentRead])
def list_assignments(agent_id: Optional[int] = None, task_id: Optional[int] = None, status: Optional[str] = None):
    with get_session() as session:
        stmt = select(TaskAssignment)
        if agent_id is not None:
            stmt = stmt.where(TaskAssignment.agent_id == agent_id)
        if task_id is not None:
            stmt = stmt.where(TaskAssignment.task_id == task_id)
        if status is not None:
            stmt = stmt.where(TaskAssignment.status == status)
        assignments = session.exec(stmt.order_by(TaskAssignment.updated_at.desc())).all()
        return assignments


@router.patch("/assignments/{assignment_id}", response_model=TaskAssignmentRead)
def update_assignment(assignment_id: int, payload: TaskAssignmentUpdate):
    with get_session() as session:
        assignment = session.get(TaskAssignment, assignment_id)
        if not assignment:
            raise HTTPException(status_code=404, detail="Assignment not found")
        for k, v in payload.dict(exclude_unset=True).items():
            setattr(assignment, k, v)
        assignment.updated_at = datetime.utcnow()
        session.add(assignment)
        session.commit()
        session.refresh(assignment)
        return assignment


@router.post("/{agent_id}/message", response_model=AgentMessageRead)
def send_message(agent_id: int, payload: AgentMessageSend, request: Request):
    with get_session() as session:
        from_agent = session.get(Agent, agent_id)
        if not from_agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        to_agent = session.get(Agent, payload.to_agent_id)
        if not to_agent:
            raise HTTPException(status_code=404, detail="Recipient agent not found")
        message = AgentMessage(
            from_agent_id=agent_id,
            to_agent_id=payload.to_agent_id,
            content=payload.content,
            message_type=payload.message_type,
            task_id=payload.task_id,
            read=False,
        )
        session.add(message)
        session.commit()
        session.refresh(message)

        mqtt_service = getattr(request.app.state, "mqtt", None)
        if mqtt_service is not None:
            mqtt_service.publish(
                f"agents/{payload.to_agent_id}/inbox",
                {
                    "id": message.id,
                    "from_agent_id": message.from_agent_id,
                    "to_agent_id": message.to_agent_id,
                    "content": message.content,
                    "message_type": message.message_type,
                    "task_id": message.task_id,
                    "timestamp": message.created_at.isoformat() + "Z",
                },
            )
        return _message_payload(message)


@router.get("/{agent_id}/messages", response_model=list[AgentMessageRead])
def list_messages(agent_id: int):
    with get_session() as session:
        stmt = (
            select(AgentMessage)
            .where((AgentMessage.from_agent_id == agent_id) | (AgentMessage.to_agent_id == agent_id))
            .order_by(AgentMessage.created_at.desc())
        )
        messages = session.exec(stmt).all()
        return [_message_payload(message) for message in messages]


@router.get("/{agent_id}/inbox", response_model=list[AgentMessageRead])
def get_inbox(agent_id: int):
    with get_session() as session:
        stmt = (
            select(AgentMessage)
            .where(AgentMessage.to_agent_id == agent_id)
            .where(AgentMessage.read == False)  # noqa: E712
            .order_by(AgentMessage.created_at.asc())
        )
        messages = session.exec(stmt).all()
        for message in messages:
            message.read = True
            session.add(message)
        session.commit()
        return [_message_payload(message) for message in messages]

File: integrations/mc-server/src/routes/agents.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
    routing::{get, post},
    Json, Router,
};
use chrono::Utc;
use serde::Deserialize;
use sqlx::Row;
use std::sync::Arc;

use crate::{
    models::agent::{
        Agent, AgentCreate, AgentMessage, AgentSession, AgentUpdate, AssignmentCreate,
        AssignmentUpdate, MessageSend, SessionCreate, TaskAssignment,
    },
    state::AppState,
};

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/agents", get(list_agents).post(create_agent))
        .route("/agents/{agent_id}", get(get_agent).patch(update_agent))
        .route("/agents/{agent_id}/sessions", get(list_sessions).post(start_session))
        .route("/agents/{agent_id}/sessions/{session_id}/end", post(end_session))
        .route("/agents/{agent_id}/message", post(send_message))
        .route("/agents/{agent_id}/messages", get(list_messages))
        .route("/agents/{agent_id}/inbox", get(get_inbox))
        .route("/agents/assignments", get(list_assignments).post(create_assignment))
        .route("/agents/assignments/{assignment_id}", axum::routing::patch(update_assignment))
}

fn row_to_agent(row: &sqlx::postgres::PgRow) -> Agent {
    Agent {
        id: row.get("id"),
        name: row.get("name"),
        capabilities: row.get("capabilities"),
        status: row.get("status"),
        metadata: row.get("metadata"),
        created_at: row.get("created_at"),
        updated_at: row.get("updated_at"),
    }
}

fn not_found(msg: &str) -> axum::response::Response {
    (StatusCode::NOT_FOUND, Json(serde_json::json!({"detail": msg}))).into_response()
}

#[derive(Deserialize)]
struct ListQuery {
    status: Option<String>,
    limit: Option<i64>,
    agent_id: Option<i32>,
    task_id: Option<i32>,
}

// ── Agents ────────────────────────────────────────────────────────────────────

async fn list_agents(
    State(state): State<Arc<AppState>>,
    Query(q): Query<ListQuery>,
) -> impl IntoResponse {
    let limit = q.limit.unwrap_or(100).min(500);
    let rows = if let Some(s) = &q.status {
        sqlx::query("SELECT * FROM agent WHERE status=$1 ORDER BY updated_at DESC LIMIT $2")
            .bind(s).bind(limit).fetch_all(&state.db).await
    } else {
        sqlx::query("SELECT * FROM agent ORDER BY updated_at DESC LIMIT $1")
            .bind(limit).fetch_all(&state.db).await
    };
    match rows {
        Ok(rows) => Json(rows.iter().map(row_to_agent).collect::<Vec<_>>()).into_response(),
        Err(e) => { tracing::error!("list_agents: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn create_agent(
    State(state): State<Arc<AppState>>,
    Json(payload): Json<AgentCreate>,
) -> impl IntoResponse {
    let now = Utc::now().naive_utc();
    let result = sqlx::query(
        "INSERT INTO agent (name, capabilities, status, metadata, created_at, updated_at) \
         VALUES ($1,$2,$3,$4,$5,$5) RETURNING *"
    )
    .bind(&payload.name).bind(&payload.capabilities).bind(&payload.status)
    .bind(&payload.metadata).bind(now)
    .fetch_one(&state.db).await;

    match result {
        Ok(row) => (StatusCode::OK, Json(row_to_agent(&row))).into_response(),
        Err(e) if e.to_string().contains("unique") || e.to_string().contains("duplicate") => {
            (StatusCode::CONFLICT, Json(serde_json::json!({"detail": "Agent name already exists"}))).into_response()
        }
        Err(e) => { tracing::error!("create_agent: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn get_agent(
    State(state): State<Arc<AppState>>,
    Path(agent_id): Path<i32>,
) -> impl IntoResponse {
    match sqlx::query("SELECT * FROM agent WHERE id=$1").bind(agent_id).fetch_optional(&state.db).await {
        Ok(Some(row)) => Json(row_to_agent(&row)).into_response(),
        Ok(None) => not_found("Agent not found"),
        Err(e) => { tracing::error!("get_agent: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn update_agent(
    State(state): State<Arc<AppState>>,
    Path(agent_id): Path<i32>,
    Json(payload): Json<AgentUpdate>,
) -> impl IntoResponse {
    let existing = sqlx::query("SELECT * FROM agent WHERE id=$1")
        .bind(agent_id).fetch_optional(&state.db).await;
    let agent = match existing {
        Ok(Some(r)) => row_to_agent(&r),
        Ok(None) => return not_found("Agent not found"),
        Err(e) => { tracing::error!("update_agent fetch: {e}"); return StatusCode::INTERNAL_SERVER_ERROR.into_response(); }
    };

    let name         = payload.name.unwrap_or(agent.name);
    let capabilities = payload.capabilities.unwrap_or(agent.capabilities);
    let status       = payload.status.unwrap_or(agent.status);
    let metadata     = payload.metadata.unwrap_or(agent.metadata);
    let now = Utc::now().naive_utc();

    match sqlx::query(
        "UPDATE agent SET name=$2, capabilities=$3, status=$4, metadata=$5, updated_at=$6 WHERE id=$1 RETURNING *"
    )
    .bind(agent_id).bind(&name).bind(&capabilities).bind(&status).bind(&metadata).bind(now)
    .fetch_one(&state.db).await {
        Ok(row) => Json(row_to_agent(&row)).into_response(),
        Err(e) => { tracing::error!("update_agent: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

// ── Sessions ──────────────────────────────────────────────────────────────────

async fn list_sessions(
    State(state): State<Arc<AppState>>,
    Path(agent_id): Path<i32>,
    Query(q): Query<ListQuery>,
) -> impl IntoResponse {
    let limit = q.limit.unwrap_or(50).min(200);
    match sqlx::query_as::<_, AgentSession>(
        "SELECT id, agent_id, context, started_at, ended_at, claude_session_id, end_reason, audit_log \
         FROM agentsession WHERE agent_id=$1 ORDER BY started_at DESC LIMIT $2"
    )
    .bind(agent_id).bind(limit).fetch_all(&state.db).await {
        Ok(rows) => Json(rows).into_response(),
        Err(e) => { tracing::error!("list_sessions: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn start_session(
    State(state): State<Arc<AppState>>,
    Path(agent_id): Path<i32>,
    Json(payload): Json<SessionCreate>,
) -> impl IntoResponse {
    let agent_exists: Option<i32> = sqlx::query_scalar("SELECT 1 FROM agent WHERE id=$1")
        .bind(agent_id).fetch_optional(&state.db).await.unwrap_or(None);
    if agent_exists.is_none() { return not_found("Agent not found"); }

    let now = Utc::now().naive_utc();
    let _ = sqlx::query("UPDATE agent SET status='online', updated_at=$2 WHERE id=$1")
        .bind(agent_id).bind(now).execute(&state.db).await;

    match sqlx::query_as::<_, AgentSession>(
        "INSERT INTO agentsession (agent_id, context, started_at) VALUES ($1,$2,$3) \
         RETURNING id, agent_id, context, started_at, ended_at, claude_session_id, end_reason, audit_log"
    )
    .bind(agent_id).bind(&payload.context).bind(now)
    .fetch_one(&state.db).await {
        Ok(s) => (StatusCode::OK, Json(s)).into_response(),
        Err(e) => { tracing::error!("start_session: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn end_session(
    State(state): State<Arc<AppState>>,
    Path((agent_id, session_id)): Path<(i32, i32)>,
) -> impl IntoResponse {
    let agent_exists: Option<i32> = sqlx::query_scalar("SELECT 1 FROM agent WHERE id=$1")
        .bind(agent_id).fetch_optional(&state.db).await.unwrap_or(None);
    if agent_exists.is_none() { return not_found("Agent not found"); }

    let now = Utc::now().naive_utc();
    let _ = sqlx::query("UPDATE agent SET status='offline', updated_at=$2 WHERE id=$1")
        .bind(agent_id).bind(now).execute(&state.db).await;

    match sqlx::query_as::<_, AgentSession>(
        "UPDATE agentsession SET ended_at=$3 WHERE id=$1 AND agent_id=$2 \
         RETURNING id, agent_id, context, started_at, ended_at, claude_session_id, end_reason, audit_log"
    )
    .bind(session_id).bind(agent_id).bind(now)
    .fetch_optional(&state.db).await {
        Ok(Some(s)) => Json(s).into_response(),
        Ok(None) => not_found("Session not found"),
        Err(e) => { tracing::error!("end_session: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

// ── Assignments ───────────────────────────────────────────────────────────────

async fn list_assignments(
    State(state): State<Arc<AppState>>,
    Query(q): Query<ListQuery>,
) -> impl IntoResponse {
    let limit = q.limit.unwrap_or(100).min(500);
    let mut sql = "SELECT * FROM taskassignment WHERE 1=1".to_string();
    let mut params: Vec<String> = vec![];
    if q.agent_id.is_some() { params.push(format!("agent_id=${}", params.len() + 2)); }
    if q.task_id.is_some()  { params.push(format!("task_id=${}", params.len() + 2)); }
    if q.status.is_some()   { params.push(format!("status=${}", params.len() + 2)); }
    if !params.is_empty() { sql = format!("{} AND {}", sql, params.join(" AND ")); }
    sql = format!("{} ORDER BY updated_at DESC LIMIT $1", sql);

    // Simple approach: fetch all then filter in Rust (acceptable for small datasets)
    match sqlx::query_as::<_, TaskAssignment>("SELECT * FROM taskassignment ORDER BY updated_at DESC LIMIT $1")
        .bind(limit).fetch_all(&state.db).await {
        Ok(mut rows) => {
            if let Some(aid) = q.agent_id { rows.retain(|r| r.agent_id == aid); }
            if let Some(tid) = q.task_id  { rows.retain(|r| r.task_id == tid); }
            if let Some(s) = &q.status    { rows.retain(|r| r.status == *s); }
            Json(rows).into_response()
        }
        Err(e) => { tracing::error!("list_assignments: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn create_assignment(
    State(state): State<Arc<AppState>>,
    Json(payload): Json<AssignmentCreate>,
) -> impl IntoResponse {
    let now = Utc::now().naive_utc();
    match sqlx::query_as::<_, TaskAssignment>(
        "INSERT INTO taskassignment (task_id, agent_id, status, created_at, updated_at) \
         VALUES ($1,$2,$3,$4,$4) RETURNING *"
    )
    .bind(payload.task_id).bind(payload.agent_id).bind(&payload.status).bind(now)
    .fetch_one(&state.db).await {
        Ok(a) => (StatusCode::OK, Json(a)).into_response(),
        Err(e) => { tracing::error!("create_assignment: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn update_assignment(
    State(state): State<Arc<AppState>>,
    Path(assignment_id): Path<i32>,
    Json(payload): Json<AssignmentUpdate>,
) -> impl IntoResponse {
    let existing = sqlx::query_as::<_, TaskAssignment>("SELECT * FROM taskassignment WHERE id=$1")
        .bind(assignment_id).fetch_optional(&state.db).await;
    let a = match existing {
        Ok(Some(a)) => a,
        Ok(None) => return not_found("Assignment not found"),
        Err(e) => { tracing::error!("update_assignment: {e}"); return StatusCode::INTERNAL_SERVER_ERROR.into_response(); }
    };
    let status = payload.status.unwrap_or(a.status);
    let now = Utc::now().naive_utc();
    match sqlx::query_as::<_, TaskAssignment>(
        "UPDATE taskassignment SET status=$2, updated_at=$3 WHERE id=$1 RETURNING *"
    )
    .bind(assignment_id).bind(&status).bind(now)
    .fetch_one(&state.db).await {
        Ok(a) => Json(a).into_response(),
        Err(e) => { tracing::error!("update_assignment patch: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

// ── Messages ──────────────────────────────────────────────────────────────────

async fn send_message(
    State(state): State<Arc<AppState>>,
    Path(agent_id): Path<i32>,
    Json(payload): Json<MessageSend>,
) -> impl IntoResponse {
    let from_exists: Option<i32> = sqlx::query_scalar("SELECT 1 FROM agent WHERE id=$1")
        .bind(agent_id).fetch_optional(&state.db).await.unwrap_or(None);
    if from_exists.is_none() { return not_found("Agent not found"); }
    let to_exists: Option<i32> = sqlx::query_scalar("SELECT 1 FROM agent WHERE id=$1")
        .bind(payload.to_agent_id).fetch_optional(&state.db).await.unwrap_or(None);
    if to_exists.is_none() { return not_found("Recipient agent not found"); }

    let now = Utc::now().naive_utc();
    match sqlx::query_as::<_, AgentMessage>(
        "INSERT INTO agentmessage (from_agent_id, to_agent_id, content, message_type, task_id, read, created_at) \
         VALUES ($1,$2,$3,$4,$5,false,$6) RETURNING *"
    )
    .bind(agent_id).bind(payload.to_agent_id).bind(&payload.content)
    .bind(&payload.message_type).bind(payload.task_id).bind(now)
    .fetch_one(&state.db).await {
        Ok(m) => (StatusCode::OK, Json(m)).into_response(),
        Err(e) => { tracing::error!("send_message: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn list_messages(
    State(state): State<Arc<AppState>>,
    Path(agent_id): Path<i32>,
    Query(q): Query<ListQuery>,
) -> impl IntoResponse {
    let limit = q.limit.unwrap_or(50).min(200);
    match sqlx::query_as::<_, AgentMessage>(
        "SELECT * FROM agentmessage WHERE from_agent_id=$1 OR to_agent_id=$1 \
         ORDER BY created_at DESC LIMIT $2"
    )
    .bind(agent_id).bind(limit).fetch_all(&state.db).await {
        Ok(msgs) => Json(msgs).into_response(),
        Err(e) => { tracing::error!("list_messages: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}

async fn get_inbox(
    State(state): State<Arc<AppState>>,
    Path(agent_id): Path<i32>,
    Query(q): Query<ListQuery>,
) -> impl IntoResponse {
    let limit = q.limit.unwrap_or(50).min(200);
    let msgs = sqlx::query_as::<_, AgentMessage>(
        "UPDATE agentmessage SET read=true \
         WHERE id IN (SELECT id FROM agentmessage WHERE to_agent_id=$1 AND read=false \
                      ORDER BY created_at ASC LIMIT $2) \
         RETURNING *"
    )
    .bind(agent_id).bind(limit).fetch_all(&state.db).await;

    match msgs {
        Ok(m) => Json(m).into_response(),
        Err(e) => { tracing::error!("get_inbox: {e}"); StatusCode::INTERNAL_SERVER_ERROR.into_response() }
    }
}
